src/backend/distributed/planner/distributed_planner.c

```diff
@@ -2321,6 +2321,77 @@ HasUnresolvedExternParamsWalker(Node *expression, ParamListInfo boundParams)
 }
 
 
+/*
+ * IsExternParamUsedInQuery returns true if the passed in paramId
+ * is used in the query, false otherwise.
+ */
+bool
+GetParamsUsedInQuery(Node *expression, List **paramListInQuery)
+{
+	if (expression == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(expression, Param))
+	{
+		Param *param = (Param *) expression;
+		int paramId = param->paramid;
+
+		/* only care about user supplied parameters */
+		if (param->paramkind != PARAM_EXTERN)
+		{
+			return false;
+		}
+
+		/* Found a parameter, add it to the list and continue */
+		*paramListInQuery = lappend_int(*paramListInQuery, paramId);
+
+		/* Continue searching */
+		return false;
+	}
+
+	/* keep traversing */
+	if (IsA(expression, Query))
+	{
+		return query_tree_walker((Query *) expression,
+								 GetParamsUsedInQuery,
+								 paramListInQuery,
+								 0);
+	}
+	else
+	{
+		return expression_tree_walker(expression,
+									  GetParamsUsedInQuery,
+									  paramListInQuery);
+	}
+}
+
+
+/*
+ * MarkUnreferencedExternParams marks parameter's type to zero if the
+ * parameter is not used in the query.
+ */
+void
+MarkUnreferencedExternParams(Node *expression, ParamListInfo boundParams)
+{
+	int parameterCount = boundParams->numParams;
+	List *paramListInQuery = NIL;
+
+	/* Fetch all parameters used in the query */
+	GetParamsUsedInQuery(expression, &paramListInQuery);
+
+	/* Check for any missing parameters */
+	for (int parameterNum = 1; parameterNum <= parameterCount; parameterNum++)
+	{
+		if (!list_member_int(paramListInQuery, parameterNum))
```

Comment:
Can't we just iterate over paramListInQuery instead of visiting all parameter numbers?

wondering if it's useful to use a Bitmapset here

Reply:
Done.